src/neo4j_genai/retrievers/text2cypher.py

```diff
@@ -0,0 +1,109 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Optional
+
+import neo4j
+from neo4j.exceptions import CypherSyntaxError
+from pydantic import ValidationError
+
+from neo4j_genai.llm import LLM
+from neo4j_genai.prompts import TEXT2CYPHER_PROMPT
+from neo4j_genai.retrievers.base import Retriever
+from neo4j_genai.schema import get_schema
+from neo4j_genai.types import (
+    LLMModel,
+    Neo4jDriverModel,
+    Neo4jSchemaModel,
+    Text2CypherRetrieverModel,
+    Text2CypherSearchModel,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Text2CypherRetriever(Retriever):
+    """
+    Allows for the retrieval of records from a Neo4j database using natural language.
+    Converts a user's natural language query to a Cypher query using an LLM,
+    then retrieves records from a Neo4j database using the generated Cypher query
+    """
+
+    def __init__(
+        self, driver: neo4j.Driver, llm: LLM, neo4j_schema: Optional[str] = None
+    ) -> None:
+        try:
+            driver_model = Neo4jDriverModel(driver=driver)
+            llm_model = LLMModel(llm=llm)
+            neo4j_schema_model = (
+                Neo4jSchemaModel(neo4j_schema=neo4j_schema) if neo4j_schema else None
+            )
+            validated_data = Text2CypherRetrieverModel(
+                driver_model=driver_model,
+                llm_model=llm_model,
+                neo4j_schema_model=neo4j_schema_model,
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        super().__init__(validated_data.driver_model.driver)
+        self.llm = validated_data.llm_model.llm
+        self.neo4j_schema = (
+            validated_data.neo4j_schema_model.neo4j_schema
+            if validated_data.neo4j_schema_model
+            else get_schema(validated_data.driver_model.driver)
+        )
+
+    def search(
+        self, query_text: str, examples: Optional[list[str]] = None
+    ) -> list[neo4j.Record]:
+        """Converts query_text to a Cypher query using an LLM.
+           Retrieve records from a Neo4j database using the generated Cypher query.
+
+        Args:
+            query_text (str): The natural language query used to search the Neo4j database.
+            examples (Optional[list[str], optional): Optional user input/query pairs for the LLM to use as examples.
+
+        Raises:
+            ValidationError: If validation of the input arguments fail.
```

Comment:
Should add `CypherSyntaxError` can also be raised. We should have this at least until we implement exceptions

Reply:
Done!